File: 2024 April POTD/Optimal Strategy For A Game.java
class solve
{
    //Function to find the maximum possible amount of money we can win.
    static long countMaximum(int n, int arr[])
    {
        // Your code here
        long[][] dp = new long[n][n];
        for(long[]e:dp){
            Arrays.fill(e,-1);
        }
        return helper(arr,dp,0,arr.length-1,0);
    }
    
    static long helper(int[] arr,long[][] dp,int l,int r,int p){
        if(l>r)
            return 0;
        if(dp[l][r]!=-1)
            return dp[l][r];
        long total = 0;
        long res = p == 1 ? Long.MAX_VALUE : Long.MIN_VALUE;
        total = arr[l];
        if(p==0){
            res = Math.max(res,total+helper(arr,dp,l+1,r,1));
        }
        else
            res = Math.min(res,helper(arr,dp,l+1,r,0));
        total = arr[r];
        if(p==0){
            res = Math.max(res,total+helper(arr,dp,l,r-1,1));
        }
        else
            res = Math.min(res,helper(arr,dp,l,r-1,0));
        return dp[l][r] = res;
    }
}

File: 2025 December POTD/Optimal Strategy For A Game.cpp
class Solution {
  public:
  
    int maximumAmount(vector<int> &arr) {
        // code here
        int n = arr.size();
        vector<vector<long long>> dp(n, vector<long long>(n, 0));

        for (int g = 0; g < n; g++) {  // gap
            for (int i = 0, j = g; j < n; i++, j++) {
                if (g == 0)
                    dp[i][j] = arr[i];
                else if (g == 1)
                    dp[i][j] = max(arr[i], arr[j]);
                else {
                    long long val1 = arr[i] + min(dp[i+2][j], dp[i+1][j-1]);
                    long long val2 = arr[j] + min(dp[i+1][j-1], dp[i][j-2]);
                    dp[i][j] = max(val1, val2);
                }
            }
        }
    
        return dp[0][n-1];
    }
};
